### PR TITLE
Set up authentication with JWT and OAuth2

### PR DESCRIPTION
**Pull request by vvkin**

### CODE REVIEW

File: backend/db/schema.sql [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"userId" serial PRIMARY KEY,
+	"username" varchar(120) NOT NULL UNIQUE,
+	"email" varchar(255) NOT NULL UNIQUE,
+	"fullName" varchar(255),
+	"avatarPath" varchar(255),
+	"location" varchar(80),
+	"company" varchar(120),
+	"website" varchar(255),
+	"bio" text,
+	"isAdmin" boolean DEFAULT FALSE,
```
Comment (nosovk) [inline backend/db/schema.sql:11]: better use role based permission system, far more flexible then isAdmin

File: backend/db/schema.sql [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+);
+
+CREATE TABLE "RefreshSessions" (
+	"tokenId" bigserial PRIMARY KEY,
+	"userId" int NOT NULL REFERENCES "Users" ON DELETE CASCADE,
+	"refreshToken" uuid NOT NULL UNIQUE,
+	"userAgent" varchar(255) NOT NULL,
+	"fingerprint" varchar(200) NOT NULL,
+	"expiresIn" bigint NOT NULL,
+	"createdAt" timestamp with time zone NOT NULL DEFAULT now()
```
Comment (nosovk) [inline backend/db/schema.sql]: here we have `timestamp with time zone`, in other places `timestamp` - they both are correct, but not mix of them

File: backend/src/app.js [hunk 3]
```diff
@@ -3,9 +3,16 @@
 const AutoLoad = require('fastify-autoload');
 const path = require('path');
 
+require('dotenv').config();
+const { DOMAIN, FRONTEND_PORT } = process.env;
+
 module.exports = async function (fastify, opts) {
+  fastify.register(require('fastify-cors'), {
```
Comment (nosovk) [inline backend/src/app.js]: inline require should trigger linter, anyway it's not very easy to support where deps inlined somewhere in logic

File: backend/src/app.js [hunk 4]
```diff
@@ -3,9 +3,16 @@
...
+    origin: `http://${DOMAIN}:${FRONTEND_PORT}`,
```
Comment (nosovk) [inline backend/src/app.js]: it surely wong, in production you will have https

File: backend/src/helpers/github-email.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+const { Forbidden } = require('http-errors');
+const axios = require('axios').default;
```
Comment (nosovk) [inline backend/src/helpers/github-email.js]: probably node-fetch currently is typical package to made requests

File: frontend/src/components/github-button.jsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    const search = toQuery(githubOauth2Params);
+    new PopupWindow('github-auth', `${githubOauth2Url}?${search}`)
+      .then(response => onSuccess(response))
+      .catch(error => onFailure(error));
+  };
+
+  return (
+    <ButtonBody onClick={handleClick}>
+      {buttonText}
+      <svg aria-hidden="true" fill="currentColor" viewBox="0 0 20 20">
```
Comment (nosovk) [inline frontend/src/components/github-button.jsx]: better use inline svg components, but manual inline is very troublesome to support

File: frontend/src/helpers/get-fingerprint.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+          localStorage: true,
+          adBlock: true,
+          screenResolution: true,
+          availableScreenResolution: true,
+          enumerateDevices: true,
+          pixelRatio: true,
+          doNotTrack: true,
+          preprocessor: (key, value) => {
+            if (key === 'userAgent') {
+              const parser = new UAParser(value);
```
Comment (nosovk) [inline frontend/src/helpers/get-fingerprint.js]: why you need this?

File: frontend/src/helpers/url-utils.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+export const toParams = query => {
```
Comment (nosovk) [inline frontend/src/helpers/url-utils.js]: isn't there JS native methods for that?

File: frontend/src/pages/register/email-form.jsx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import {
+  AuthFormContent,
+  AuthInput,
+  AuthSignUpButton,
+} from '../../components/styled/auth';
+
+export const EmailForm = ({ state, updateState, updateStep, setError }) => {
+  const handleLocalRegister = event => {
+    event.preventDefault();
+    state.password.match('^(?=.*[A-Za-z])(?=.*d)[A-Za-z\\d]{8,}$')
```
Comment (nosovk) [inline frontend/src/pages/register/email-form.jsx]: all regexp should be placed in constant and have propper description

File: frontend/src/helpers/get-fingerprint.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import Fingerprint2 from 'fingerprintjs2';
+import UAParser from 'ua-parser-js';
+
+export const getFingerprint = () => {
+  return new Promise((resolve, reject) => {
+    async function getHash() {
```
Comment (nosovk) [inline frontend/src/helpers/get-fingerprint.js]: FP shouldn't be calculated on each invocation, it should persist on local storage or at least session storage

File: backend/db/schema.sql (hunk 1 above)
Comment (vvkin) [reply backend/db/schema.sql:11]: we'll add it later

File: frontend/src/helpers/url-utils.js (hunk 8 above)
Comment (vvkin) [reply frontend/src/helpers/url-utils.js]: replaced with URLSearchParams

File: frontend/src/helpers/get-fingerprint.js (hunk 10 above)
Comment (vvkin) [reply frontend/src/helpers/get-fingerprint.js]: the fingerprint is now calculated only if it is not present in the localStorage

File: frontend/src/helpers/get-fingerprint.js (hunk 7 above)
Comment (vvkin) [reply frontend/src/helpers/get-fingerprint.js]: fixed

**Pull request merged**